[core][nextjs] Investigate client/server side context id issues

MenKNas opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

- [ ] This PR follows the [Contribution Guide](https://github.com/Sitecore/content-sdk/blob/dev/CONTRIBUTING.md)
- [x] Changelog updated
- [ ] Upgrade guide entry added

## Description / Motivation
This PR standardizes the way we handle `contextId` and `clientContextId` across the content-sdk. It removes previous fallbacks and handles all possible scenarios for client/server context IDs, whether they are provided or not. After the change, the expected behavior now is the following:

- Only Client Id provided-> App fails
- Only Server Id provided-> Works with warning (warning about graphql and client side things that might not work properly)
- Both provided -> Works with no warnings/errors
- None provided -> App fails

## Testing Details
- [x] Unit Test Added
- [x] Manual Test/Other (Please elaborate)

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

MenKNas committed:
[core][nextjs] initial commit for client/server context id handling

MenKNas committed:
[core][nextjs] adjust all possible context id scenarios

MenKNas committed:
[core][nextjs]: Update tests and remove some comments

MenKNas committed:
Merge branch 'dev' into bugfix/JSS-7018-investigate-context-id-issues

MenKNas committed:
[core][nextjs]: minor comment updates

MenKNas committed:
[core][nextjs]: update CHANGELOG.md

MenKNas committed:
[core][nextjs] Fix linting issues

art-alexeyenko commented on packages/core/src/config/define-config.ts:
```diff
@@ -107,38 +107,30 @@ const resolveConfig = (base: SitecoreConfig, override: SitecoreConfigInput): Sit
+    if (!hasClientContextId) {
+      // eslint-disable-next-line no-console
+      console.warn(
+        'Warning: only a server-side `contextId` is provided. If your app makes ' +
```
Small suggestion for you to consider: the code and the error/warn messages would read better if they are formatted to be multi-line messages here
It's fine to leave the format is it was before too, up to your decision

art-alexeyenko commented on packages/core/src/client/utils.ts:
```diff
@@ -5,34 +5,42 @@ import { FetchOptions } from '../models';
   } else {
     throw new Error(
-      'Please configure and use either your sitecoreEdgeContextId, or your graphQLEndpoint and sitecoreApiKey.'
+      'GraphQL client mis-configured. Provide one of:\n' +
```
Some small gripes with wording in here:
- It's better to be more a little more prescriptive, i.e. point users to sitecore.config and/or .env where this should be configured
- api.local.{apiHost, apiKey} can be expanded to api.local.apiHost and api.local.apiKey
- Providing api.edge.clientContextId only will fail the app
- misconfigured can be written together

art-alexeyenko commented on packages/core/src/client/utils.ts:
```diff
@@ -5,34 +5,42 @@ import { FetchOptions } from '../models';
+      endpoint: getEdgeProxyContentUrl(edge.contextId, edge.edgeUrl),
     };
-  } else if (options.api?.local?.apiKey && options.api?.local?.apiHost) {
+  } else if (edge?.clientContextId) {
```
```suggestion
  } else if (isBrowser && edge?.clientContextId) {
```
to be extra consistent with the expected behavior

art-alexeyenko commented on packages/core/src/client/utils.ts:
Please also cleanup comments related to contextId and clientContextId in models.ts

art-alexeyenko reviewed (changes requested):
Good stuff, but a few non-obvious blind spots missed. Some listed in the comments, and some I'll add below:
- Please check contextId usage in `ComponentLayoutService`. Currently it is only used on server side, but let's make it complaint with the overall philoshophy and make is use clientContextId as fallback
- make sure Form.tsx in `react` package uses client context ID
- make sure SitecoreProvider context can use client ID
- I'm not sure if it's possible for `getHeadLinks` in `sitecoreClient` to get different result between server and client if context IDs differ. Please check and adjust implementation depending on the answer:
  - if expected, use clientContextId in `getHeadLinks`
  - if not expected, make sure  headlinks only render on server and no hydration errors occur

MenKNas committed:
[core][nextjs] address PR comments

MenKNas committed:
[core][nextjs] fix build issues

MenKNas committed:
[core][nextjs] fix failing tests

MenKNas committed:
[core][nextjs] adjust warning/error messages

MenKNas commented:
Hey @art-alexeyenko, thanks for the valuable feedback! It really helped me understand some parts of the codebase in greater depth. I’ve addressed all the points and added a few guardrails for future edge cases, so let me know if anything else pops up.

Regarding the last item:

After investigating this, my conclusion is that `getHeadLinks` reuses the serialized `contextId` on both server and client, so SSR and hydration stay in sync. Swapping to `clientContextId` only on the client would introduce differing URLs and risk a mismatch, so no change is required there. Of course, feel free to provide any further insight.

art-alexeyenko commented on packages/core/src/editing/component-layout-service.ts:
```diff
@@ -1,75 +1,43 @@
-   * mode to be used for rendering the component
-   */
-  mode?: DesignLibraryMode;
+  itemId: string; // context item ID
```
why were tsdoc comments removed?

art-alexeyenko commented on packages/core/src/editing/component-layout-service.ts:
```diff
@@ -80,29 +48,38 @@ export class ComponentLayoutService {
-      .catch((error) => {
-        if (error.response?.status === 404) {
-          return error.response.data;
+      .then((r) => r.data)
```
are these abbreviations neccessary?

art-alexeyenko commented on packages/core/src/editing/component-layout-service.ts:
```diff
@@ -80,29 +48,38 @@ export class ComponentLayoutService {
+
+    // Choose the correct Edge ID per environment
+    const sitecoreContextId =
+      this.config.contextId ?? (isBrowser ? this.config.clientContextId : undefined);
```
this can just be simplified to `this.config.contextId || this.config.clientContextId` . Client id can be used as fallback.
Alternatively, you can introduce a helper method that retrieves context id depending on server/browser context, to keep behavior consistent across the board

art-alexeyenko commented on packages/react/src/components/Form.tsx:
```diff
@@ -16,23 +15,13 @@ export const mockFormModule = (formModule: any) => {
-    /**
-     * The unique identifier of the rendering.
-     */
+    FormId: string; // Sitecore Form ID to render
```
tsdoc comments should not be affected

art-alexeyenko commented on packages/react/src/components/Form.tsx:
```diff
@@ -48,7 +37,10 @@ export const Form = ({ params, rendering }: FormProps) => {
     if (!content) {
-      loadForm(context.api?.edge?.contextId, params.FormId, context.api?.edge?.edgeUrl)
+      // In the browser prefer clientContextId; fall back to contextId if it exists
+      const edgeId = context.api?.edge?.clientContextId ?? context.api?.edge?.contextId;
```
client context ID should be used only - form is rendered client side

art-alexeyenko commented on packages/react/src/components/SitecoreProvider.tsx:
```diff
@@ -7,35 +7,18 @@ import { constants } from '@sitecore-content-sdk/core';
-   * @param {SitecoreProviderPageContext | LayoutServiceData} value New page context value.
-   * @returns {void}
-   */
   setContext: (value: SitecoreProviderPageContext | LayoutServiceData) => void;
```
tsdoc comments should stay

art-alexeyenko replied:
Same thing for comments for pageContext and api fields down below

art-alexeyenko commented on packages/react/src/components/SitecoreProvider.tsx:
```diff
@@ -63,9 +44,13 @@ export class SitecoreProvider extends React.Component<
     let api = props.api;
-
-    if (props.api?.edge?.contextId && !props.api?.edge?.edgeUrl) {
+    if (
```
```suggestion
    if ((props.api?.edge?.contextId || props.api?.edge?.clientContextId) && !props.api?.edge?.edgeUrl)
```

art-alexeyenko commented on packages/react/src/components/SitecoreProvider.tsx:
```diff
@@ -97,23 +80,15 @@ export class SitecoreProvider extends React.Component<
   componentDidUpdate(prevProps: SitecoreProviderProps) {
-    // In case if somebody will manage SitecoreProvider state by passing fresh `layoutData` prop
-    // instead of using `updateContext`
+    // Refresh context if a new layoutData object is passed in
```
I think the older comment has a bit more context

art-alexeyenko commented on packages/react/src/components/SitecoreProvider.tsx:
```diff
@@ -97,23 +80,15 @@ export class SitecoreProvider extends React.Component<
-   * Update context state. Value can be @type {LayoutServiceData} which will be automatically transformed
-   * or you can provide exact @type {SitecoreProviderPageContext}
-   * @param {SitecoreProviderPageContext | LayoutServiceData} value New context value
-   */
```
comment should stay

art-alexeyenko commented on packages/react/src/components/SitecoreProvider.tsx:
```diff
@@ -97,23 +80,15 @@ export class SitecoreProvider extends React.Component<
   setContext = (value: SitecoreProviderPageContext | LayoutServiceData) => {
     this.setState({
-      pageContext: value.sitecore
+      pageContext: (value as any).sitecore
```
we shouldn't use `any`

art-alexeyenko commented on packages/core/src/config/models.ts:
```diff
@@ -23,153 +23,175 @@ export type DeepPartial<T> = {
-      // for now contextID will take the role of both server and client IDs
       /**
-       * A unified identifier used to connect and retrieve data from XM Cloud instance
+       * **Server-side Edge context ID** used for SSR and API-route requests.
```
Let's leave the phrasing `A unified identifier` for both server and client IDs

art-alexeyenko commented on packages/core/src/config/models.ts:
```diff
@@ -23,153 +23,175 @@ export type DeepPartial<T> = {
       /**
-       * Optional identifier used to connect and retrieve data from XM Cloud instance in client-side functionality
+       * **Browser-side Edge context ID** used for client-side GraphQL calls.
+       * Required alongside `contextId`; supplying only this ID will cause
```
it is an optional setting

art-alexeyenko commented on packages/core/src/config/models.ts:
```diff
@@ -23,153 +23,175 @@ export type DeepPartial<T> = {
+
       /**
-       * XM Cloud endpoint that the app will communicate and retrieve data from
+       * Custom XM Cloud endpoint the app communicates with.
```
I wouldn't say it is custom. It's just the URL to contact, with default value used when not set explicitly

art-alexeyenko commented on packages/core/src/config/models.ts:
```diff
@@ -23,153 +23,175 @@ export type DeepPartial<T> = {
     /**
-     * Retry strategy for the client. By default, uses exponential
-     * back-off factor of 2 for codes 429, 502, 503, 504, 520, 521, 522, 523, 524.
+     * Retry strategy for the client.
```
this comment should not be abridged

art-alexeyenko commented on packages/core/src/config/models.ts:
```diff
@@ -23,153 +23,175 @@ export type DeepPartial<T> = {
   dictionary?: {
     /**
-     * configure local memory caching for Dictionary Service requests
+     * Configure local-memory caching for Dictionary Service requests
```
no need for a dash

art-alexeyenko reviewed (commented):
There are some more adjustments needed, mainly around variable naming and comments

MenKNas committed:
[core][nextjs] address PR comments

MenKNas committed:
[core][nextjs]: fix failing test

MenKNas committed:
[core][nextjs]: restore js docs

art-alexeyenko commented on packages/core/src/config/models.ts:
```diff
@@ -23,153 +23,174 @@ export type DeepPartial<T> = {
+     * (excluding the fields part).
+     * @param siteName  The site name
+     * @param itemPath  Full path to the Sitecore item/route
+     * @param locale    Item/route language
```
please revert these changes

art-alexeyenko commented on packages/core/src/config/models.ts:
```diff
@@ -179,89 +200,87 @@ export type SitecoreConfigInput = {
-   * Function to generate the component file contents based on the component name.
-   * @param componentName - The name of the component.
-   * @returns The generated content as a string.
+   * Function to generate the component file contents
```
please revert this too

art-alexeyenko commented on packages/core/src/config/models.ts:
```diff
@@ -179,89 +200,87 @@ export type SitecoreConfigInput = {
-   * Optional function to get the next steps to be shown by the cli after generating the component.
-   * @param componentOutputPath - The output path of the generated component.
-   * @returns An array of strings representing the next steps.
+   * Optional function to return the next steps shown after generating the component
```
please revert this too

art-alexeyenko commented on packages/core/src/editing/component-layout-service.ts:
```diff
@@ -114,11 +123,7 @@ export class ComponentLayoutService {
-   * @param {ComponentLayoutRequestParams} params - The parameters for the request
-   * @returns {string} The fetch URL for the component data
-   */
+  /** Build the HTTP URL for the partial-layout endpoint */
```
revert this change

art-alexeyenko commented on packages/react/src/components/Form.tsx:
```diff
@@ -6,8 +6,7 @@ import { useSitecore } from '../enhancers/withSitecore';
 /**
- * Mock function to replace the form module functions for `testing` purposes.
- * @param {any} formModule - The form module to mock
+ * Replace the form module functions for testing purposes.
```
revert this change

art-alexeyenko commented on packages/react/src/components/Form.tsx:
```diff
@@ -16,18 +15,17 @@ export const mockFormModule = (formModule: any) => {
   params: {
     /**
-     * The ID of the Sitecore Form to render.
+     * The unique identifier of the rendering.
```
revert this change too

art-alexeyenko commented on packages/react/src/components/SitecoreProvider.tsx:
```diff
@@ -101,8 +88,6 @@ export class SitecoreProvider extends React.Component<
     if (!fastDeepEqual(prevProps.layoutData, this.props.layoutData)) {
       this.setContext(this.props.layoutData);
-
-      return;
```
let's leave the return statement for descriptive purposes

art-alexeyenko commented on packages/core/src/client/utils.ts:
```diff
@@ -5,34 +5,43 @@ import { FetchOptions } from '../models';
- * @param {GraphQLClientOptions} options content sdk config
- * @returns GraphQLRequestClientFactory instance
- */
 export const createGraphQLClientFactory = (options: GraphQLClientOptions) => {
```
tsdoc comment should remain

art-alexeyenko commented on packages/core/src/config/models.ts:
```diff
@@ -23,153 +23,174 @@ export type DeepPartial<T> = {
+
   /**
-   * Your default site name. When using the multisite feature this variable defines the fallback site.
+   * Default site name. When using multisite, this is the fallback site.
```
Please retain `this variable defines the fallback site` wording, as our multisite logic is somewhat specific like that

art-alexeyenko reviewed (approved):
I'm approving this, as only the comment adjustments remain. Feel free to merge once you address them.
Overall, please be mindful of changes when adjusting code with Cursor. It would be better to mainly apply changes relevant to the task to not lose focus during review and implementation.

MenKNas committed:
[core][nextjs]: address PR comments
